날씨 정보 (전) [STEP 2] 산체스

sanches37 opened the pull request:
@jungseungyeo

안녕하세요 린생! 
PR을 보냅니다. 잘 부탁드려요~ 😆

### 고민한점
1. 위도 경도 정보를 얻기 위해 StartUpdating, startMonitoringSignificantLocationChanges,  requestLocation 중 어떤 것을 선택할지를 고민했습니다. 지속적으로 정보를 받아야 된다면, StartUpdating, startMonitoringSignificantLocationChanges 를 사용해야 된다고 생각했지만, step3 에서 Refresh Control 을 구현하라는 내용이 있어서, Refresh Control 을 이용한 최신화 및 앱을 키거나 백그라운드에서 화면을 다시 불러올 때 정도만 정보를 최신화 시키면 된다고 판단하여 requestLocation 을 이용해서 한번만 불러오는 방법을 선택했습니다. 아직도 기존 프로젝트에서 지속적으로 정보를 받아오는게 나은지, 지금처럼 필요한 경우에만 정보요청을 하는게 나은지 고민입니다. 린생의 의견이 궁금합니다.


2.
``` swift
private func processWeatherImpormation() {
        locationManager.getUserLocation { [weak self] location in
            guard let self = self else { return }
            let currentWeatherURL =
            WeatherURL.weatherCoordinates(latitude: location.coordinate.latitude,
                                          longitude: location.coordinate.longitude)
            let fiveDaysWeatherURL =
            WeatherURL.forecastCoordinates(latitude: location.coordinate.latitude,
                                          longitude: location.coordinate.longitude)
            self.getCurrentWeatherImpormation(request: currentWeatherURL)
            self.getFiveDaysWeatherImpormation(request: fiveDaysWeatherURL)
            self.getUserAddress()
        }
    }
```
 processWeatherImpormation 메서드에서 다른처리를 하는 메서드들을 묶은 후 ViewDidLoad 에서 호출합니다. 주소 정보를 받는 getUserAddress() 는 따로 분리하여 ViewDidLoad 에서 호출하려 했지만, 앱 구조상 getUserLocation 을 호출해서 Location 을 받아오지 않으면 getUserAddress 내 Location 이 nil 인 상황이 되어서 호출이 제대로 되지 않았습니다. 구조를 수정해볼까도 생각해봤지만, 추후 refresh control 을 사용할 때 이 메서드 하나만 다시 호출해주면 될 것 같아서 그대로 두었습니다. 


3.
``` swift
private func getCurrentWeatherImpormation(request: UrlGeneratable) {
        self.apiManager.requestAPI(
            resource: APIResource(apiURL: request)) { result in
                switch result {
                case .success(let data):
                    guard let product =
                            try? self.decodingManager.decodeJSON(
                                CurrentWeather.self, from: data) else {
                                    return
                                }
                    debugPrint(product)
                case .failure(let error):
                    self.handlerError(error)
                }
            }
    }
    
    private func getFiveDaysWeatherImpormation(request: UrlGeneratable) {
        self.apiManager.requestAPI(
            resource: APIResource(apiURL: request)) { result in
                switch result {
                case .success(let data):
                    guard let product =
                            try? self.decodingManager.decodeJSON(
                                FiveDaysWeather.self, from: data) else {
                                    return
                                }
                    debugPrint(product)
                case .failure(let error):
                    self.handlerError(error)
                }
            }
    }
```

getCurrentWeatherImpormation 와 getFiveDaysWeatherImpormation 가 동일 코드가 반복됩니다. 현재 상황에선 모델타입만 메서드로 받으면 하나의 메서드로 사용할 수 있지만, 추후 디코딩한 파일을 화면에 그릴 때 두 메서드가 달라질 걸로 생각됩니다. 달라지는 부분을 매개변수로 구분하더라도 중복코드를 줄이는게 나을까요?

sanches37 committed:
feat: LocationManager 클래스 추가

- 현재위치 획득을 위한 LocationManager 추가

sanches37 committed:
refactor: WeatherURL enum 수정

- case 가 상단에 오도록 순서 변경

sanches37 committed:
refactor: APIResource 타입 코드 수정

- components 의 항목을 String으로 변환해서 받았던 구조를 url 그대로 받을 수 있도록 수정
- queryString 이 배열에 담기지 않아서 코드 추가

sanches37 committed:
refactor: LocationManager 권한요청 방식 변경

- locationManager(didChangeAuthorization status:) 메서드를 이용해서 권한 요청하도록 변경
- viewController -> WeatherImpormationViewController 로 이름 변경
- info.plist 권한요청 추가

sanches37 committed:
feat: LocationManager 주소 전달 메서드 생성

- getUserAddress 생성
- LocationManager 위치 정리
- LocationError enum 생성

sanches37 committed:
refactor: getUserAddress 메서드 String 을 리턴하도록 수정

- 필요한 정보인 administrativeArea, thororoughfare 만 String 으로 전달할 수 있도록 구조 변경
- LocationError 타입 내 invalidAddress case 추가

sanches37 committed:
refactor: 정보 업데이트 방식 변경

-  앱의 특성에 따라 한번 호출에 적합한 requestLocation 메서드로 변경
- 정보 업데이트 코드를 getUserLocation 에서 사용했는데, didChangeAuthorization 에서 사용하도록 변경
- didFailWithError 를 추가하여 위치가 받아지지 않았을 때 error 처리 코드 추가

sanches37 committed:
refactor: LocationManager 구조변경

- locationManager.delegate = self 코드를 init 에서 하도록 수정
- 추후 화면 refresh 를 위한 getUserLocation() 에 requestLocation() 추가
- 위치정보를 받지 못했을 때 재시도를 위해서 didfailWithError 에 requestLocation() 추가

sanches37 committed:
refactor:  APiError 에 Error 프로토콜 채택이 안된 것 추가

sanches37 committed:
refactor: 에러핸들러 디버그프린트로 변경

- 기존 print 대신 debugPrint 를 사용하도록 수정

sanches37 committed:
feat: viewController 에서 날씨정보를 가져오는 코드 추가

- 주소를 가져오는 getUserAddress() 추가
- 날씨정보를 네트워크로 가져오는 getCurrentWeatherImpormation() 추가
- 날씨정보를 네트워크로 가져오는 getFiveDaysWeatherImpormation() 추가
- 날씨정보 및 주소를 한번에 최신화 할 수 있는 processWeatherImpormation() 추가

jungseungyeo reviewed (changes requested):
1. 위도 경도 정보를 얻기 위해 StartUpdating, startMonitoringSignificantLocationChanges, requestLocation 중 어떤 것을 선택할지를 고민했습니다. 지속적으로 정보를 받아야 된다면, StartUpdating, startMonitoringSignificantLocationChanges 를 사용해야 된다고 생각했지만, step3 에서 Refresh Control 을 구현하라는 내용이 있어서, Refresh Control 을 이용한 최신화 및 앱을 키거나 백그라운드에서 화면을 다시 불러올 때 정도만 정보를 최신화 시키면 된다고 판단하여 requestLocation 을 이용해서 한번만 불러오는 방법을 선택했습니다. 아직도 기존 프로젝트에서 지속적으로 정보를 받아오는게 나은지, 지금처럼 필요한 경우에만 정보요청을 하는게 나은지 고민입니다. 린생의 의견이 궁금합니다.

> 백그라운에서 동작을 하지 않게 한다고 생각하면 괜찮으거 같아요.

2.processWeatherImpormation 메서드에서 다른처리를 하는 메서드들을 묶은 후 ViewDidLoad 에서 호출합니다. 주소 정보를 받는 getUserAddress() 는 따로 분리하여 ViewDidLoad 에서 호출하려 했지만, 앱 구조상 getUserLocation 을 호출해서 Location 을 받아오지 않으면 getUserAddress 내 Location 이 nil 인 상황이 되어서 호출이 제대로 되지 않았습니다. 구조를 수정해볼까도 생각해봤지만, 추후 refresh control 을 사용할 때 이 메서드 하나만 다시 호출해주면 될 것 같아서 그대로 두었습니다.

> 넵 현재 구조는 좋아보이지 않아요 클로져 안에서 api를 호출하는 부분도 좋지 않은거 같아요.

3. getCurrentWeatherImpormation 와 getFiveDaysWeatherImpormation 가 동일 코드가 반복됩니다. 현재 상황에선 모델타입만 메서드로 받으면 하나의 메서드로 사용할 수 있지만, 추후 디코딩한 파일을 화면에 그릴 때 두 메서드가 달라질 걸로 생각됩니다. 달라지는 부분을 매개변수로 구분하더라도 중복코드를 줄이는게 나을까요?

> 함수 하나로 합치면 좋을거 같아요. 제네릭을 활요하거나 추상화를 활용하면 좋을거같아요.

산체스 코드를 보면서 많이 좋아지고 있는 것이 보여요. 여기서 우리 하나만 추가하면 좋을거 같아요. 

LocationManager의 대한 테스트 코드를 작성해주면 좋을거같아요. protocol을 활용하는 방법을 고민하고 어떻게해야 테스트가 가능한 구조가 되는지 고민해보주세요. 

구현부분에서는 너무 잘 하셨지만 테스트가 가능한 구조를 만드는다는 것에서는 많이 부족한 모습을 보여주고 있는거 같아요. 이 부분도 고려해서 코드를 수정하면 좋을거같아요.

sanches37 committed:
refactor: LocationManager 테스트 위한 코드 추가

- Unit Test 코드 구현
- MockLocationManager 구현
- 의존성 주입을 위한 LocationManagerProtocol, LocationManagerDelegate 프로토콜 구현
- 의존성 주입을 위해 LocationManager 타입 수정

sanches37 committed:
refactor: API 호출 함수 하나로 통합

- getFiveDaysWeatherImpormation, ,getCurrentWeatherImpormation -> getWeatherImpormation 로 메서드 통합

sanches37 committed:
refactor: processWeatherImpormation() 함수 내부 분리

- WeatherImpormationViewController 내 processWeatherImpormation() 메서드가 많은 작업을 하고 있어서 코드 분리
- API 호출을 통한 코드는 processWeatherImpormation() 메서드에서 처리하도록 수정
- locationManager.getUserLocation 의 호출은 getUserLocation 에서 처리하도록 수정
- processWeatherImpormation, getUserAddress 는 ViewDidAppear 에서 처리하도록 수정
- locationManager.getUserAddress 메서드 매개변수 받는 구조로 수정

sanches37 committed:
refactor: getUserLocation 메서드 weak self 제거

sanches37 committed:
refactor: didFailWithError 내 requestLocation 제거

sanches37 committed:
refactor: getUserLocation() 의 클로저에서 함수 호출하도록 변경

- 작동 불가로 인해 getUserLocation() 의 클로저에서 함수 호출하도록 변경

sanches37 committed:
refactor: 날씨정보 및 주소 프로퍼티 옵저버에서 호출하도록 변경

- currentLocation 의 didSet 에서 processWeatherImpormation(), getUserAddress() 호출하도록 변경

jungseungyeo commented on WeatherForecast/WeatherForecast/Managers/LocationManager.swift:
```diff
@@ -9,14 +9,14 @@ import Foundation
 
 class LocationManager: NSObject {
-    private let locationManager = CLLocationManager()
+    private var locationManager: LocationManagerProtocol
```
의존성 주입을 하기 위하여 protocol을 활요하셨군요 ㅎㅎ 너무 좋습니다.

jungseungyeo commented on WeatherForecast/WeatherForecast/Managers/LocationManager.swift:
```diff
@@ -9,14 +9,14 @@ import Foundation
     private var currentLocation: CLLocation?
     
-    override init() {
+    init(locationManager: LocationManagerProtocol = CLLocationManager()) {
```
너무 좋아요ㅎㅎ 제가 의도한 방법과 동일하게 작성해주셨어요 ㅎㅎ 의존성이라는 것이 무엇일까요?  알고 계실거 같지만 한번 코멘트 달아주세요 :)

jungseungyeo commented on WeatherForecast/WeatherForecast/Managers/LocationManager.swift:
```diff
@@ -9,14 +9,14 @@ import Foundation
         super.init()
-        
-        locationManager.delegate = self
+        self.locationManager.locationManagerDelegate = self
```
직접 만든 delegate를 만드셨네요 이것의 의도가 있나요?

jungseungyeo commented on WeatherForecast/WeatherForecast/Managers/LocationManager.swift:
```diff
@@ -51,7 +51,7 @@ class LocationManager: NSObject {
 }
 
-extension LocationManager: CLLocationManagerDelegate {
+extension LocationManager: LocationManagerDelegate {
```
프로토콜을 상속을 사용하셨군요 ㅎㅎ 이것이 일반 상속이랑 무엇이 다를까요?

jungseungyeo commented on WeatherForecast/WeatherForecast/Protocols/LocationProtocol.swift:
```diff
@@ -0,0 +1,35 @@
+import CoreLocation
+
+protocol LocationManagerProtocol {
+    var locationManagerDelegate: LocationManagerDelegate? { get set }
```
이 부분 질문을 주셨는데요. 사용해도 괜찮은거 같아요. 해당 구현을 위해서 필요한 부분이고 의존성을 주입하기 떄문에 괜찮으거 같아요.

jungseungyeo commented on WeatherForecast/WeatherForecastTests/WeatherForecastTests.swift:
```diff
@@ -5,28 +5,30 @@
+            XCTAssertEqual(location.coordinate.longitude,
+                           expectLocation.coordinate.longitude)
+        }
+        wait(for: [expectation], timeout: 2)
```
테스트 코드 너무 잘 작성해주셨어요 여기서 하나만 수정 하면 좋을거같아요. 클로져 안에서 결과 값을 도출하는 것 보다는 외부에서 결과를 처리하는 방식으로 수정해주세요. 크게 문제가 되거나 그런거는 아니지만 비동기 처리를 순차적으로 처리하는 것처럼 보이게 하는게 저는 좋은거 같더라고요 ㅎㅎ

휴먼 리소스를 관리 용으로 괜찮으거 같아요. 혹시 산체스 의견은 어떤가요?

jungseungyeo reviewed (commented):
산체스 테스트 코드 너무 잘 안드셨어요. 기존 코드를 보면서 테스트가 불가능한 코드에서 테스트가 가능한 코드로 변화는 것을 느꼈을거 같아요. 항상 테스트가 가능한 코드를 작성을 하는것을 의도하면 좋을거같아요.

마지막으로 코멘트에 답변만 달아주세요 :) 고생하셨습니다. 산체스!

sanches37 committed:
test: Test 코드 수정

-  클로져 안에서 결과 값을 도출하는 것을 외부에서 결과를 처리하는 방식으로 수정

jungseungyeo commented:
산체스 고생하셨습니다. 피드백 반영을 잘 해주신거 같아요. step3 진행 부탁드립니다. :)
